### PR TITLE
Refactor/issue 397

### LINKED ISSUE #397: [POS] StakeValidator.CheckKernel accesses disposed coinview
```
      PosMining.Mine threw an unhandled exception
System.ObjectDisposedException: Cannot access a disposed object.
Object name: 'DBreezeSession'.
   at Stratis.Bitcoin.Utilities.DBreezeSingleThreadSession.AssertNotDisposed()
   at Stratis.Bitcoin.Utilities.DBreezeSingleThreadSession.Execute[T](Func`1 act)
   at Stratis.Bitcoin.Features.Consensus.CoinViews.DBreezeCoinView.FetchCoinsAsync(uint256[] txIds)
   at Stratis.Bitcoin.Features.Consensus.CoinViews.CachedCoinView.<FetchCoinsAsync>d__26.MoveNext()
--- End of stack trace from previous location where exception was thrown ---
   at System.Runtime.ExceptionServices.ExceptionDispatchInfo.Throw()
   at System.Runtime.CompilerServices.TaskAwaiter.HandleNonSuccessAndDebuggerNotification(Task task)
   at System.Runtime.CompilerServices.TaskAwaiter`1.GetResult()
   at Stratis.Bitcoin.Features.Consensus.StakeValidator.CheckKernel(ContextInformation context, ChainedBlock pindexPrev, UInt32 nBits, Int64 nTime, OutPoint prevout, Int64& pBlockTime)
   at Stratis.Bitcoin.Features.Miner.PosMinting.CreateCoinStake(List`1 stakeTxes, ChainedBlock pindexBest, Block block, Int64 nSearchInterval, Int64 fees, Transaction& txNew, Key& key)
   at Stratis.Bitcoin.Features.Miner.PosMinting.SignBlock(List`1 stakeTxes, Block block, ChainedBlock pindexBest, Int64 fees)
   at Stratis.Bitcoin.Features.Miner.PosMinting.GenerateBlocks(WalletSecret walletSecret)
   at Stratis.Bitcoin.Features.Miner.PosMinting.<>c__DisplayClass33_0.<Mine>b__0(CancellationToken token)
   at Stratis.Bitcoin.Utilities.AsyncLoop.<>c__DisplayClass8_0.<<StartAsync>b__0>d.MoveNext()
```

this happened when I pressed ctrl+C and it was probably this line in the code 

```
            FetchCoinsResponse coins = this.coinView.FetchCoinsAsync(new[] { prevout.Hash }).GetAwaiter().GetResult();
``` 

that accessed coinView that was disposed at that moment.
Mining should be terminated before coinviews are disposed.

### PR DESCRIPTION
**Pull request by Aprogiena**
This removes one check for app shutdown as it seems unnecessary (as argued in #397), the second check is replaced with nicer construction, that quits the resource heavy method instead of throwing exception. This will cause logs to end in a consistent way.

The other part of #397 is fixed, so this is the last thing to close this issue, therefore:

Fix #397

### CODE REVIEW

**Comment by fassadlr** [comment]
[[POS] 1 StakeValidator.CheckKernel accesses disposed coinview](https://trello.com/c/sSLsoEkz/188-pos-1-stakevalidatorcheckkernel-accesses-disposed-coinview)

File: Stratis.Bitcoin.Features.Miner/PosMinting.cs [hunk 1]
```diff
@@ -564,7 +564,11 @@ public bool CreateCoinStake(List<StakeTx> stakeTxes, Block block, ChainedBlock c
 
                 for (uint n = 0; (n < searchInterval) && !fKernelFound; n++)
                 {
-                    this.nodeLifetime.ApplicationStopping.ThrowIfCancellationRequested();
+                    if (this.nodeLifetime.ApplicationStopping.IsCancellationRequested)
+                    {
+                        this.logger.LogTrace("(-)[SHUTDOWN]:false");
```
Comment (bokobza) [inline Stratis.Bitcoin.Features.Miner/PosMinting.cs]: how should this log be interpreted?

File: Stratis.Bitcoin.Features.Miner/PosMinting.cs [hunk 2]
```diff
@@ -636,8 +640,6 @@ public bool CreateCoinStake(List<StakeTx> stakeTxes, Block block, ChainedBlock c
             this.logger.LogTrace("Trying to reduce our staking UTXO set by adding additional inputs to coinstake transaction...");
             foreach (StakeTx stakeTx in setCoins)
             {
-                this.nodeLifetime.ApplicationStopping.ThrowIfCancellationRequested();
```
Comment (bokobza) [inline Stratis.Bitcoin.Features.Miner/PosMinting.cs]: Why was this removed from here?

File: Stratis.Bitcoin.Features.Miner/PosMinting.cs (hunk 1 above)
Comment (Aprogiena) [reply Stratis.Bitcoin.Features.Miner/PosMinting.cs]: shutdown detected, leaving whatever i'm working on

Comment (bokobza) [reply Stratis.Bitcoin.Features.Miner/PosMinting.cs]: what's the false about? Returning false?

Comment (Aprogiena) [reply Stratis.Bitcoin.Features.Miner/PosMinting.cs]: that is return value of the method

File: Stratis.Bitcoin.Features.Miner/PosMinting.cs (hunk 2 above)
Comment (Aprogiena) [reply Stratis.Bitcoin.Features.Miner/PosMinting.cs]: that was argued in #397, it is not needed here, this part of the code is quick and will complete in no time even for large miner, so i could either replace it with the construction as before or remove it, i couldn't find reason to keep it

Comment (bokobza) [reply Stratis.Bitcoin.Features.Miner/PosMinting.cs]: cool mate, thanks.
